### ISSUE TITLE
Feature Request: Add `set` method to `FirebaseListObservable`

### ISSUE THREAD

**Issue by somombo** [angular/angularfire2#910]
I have an idea that, in addition to `push`, `update`, `remove` etc ... perhaps, for completeness, we should include destructive `set` method in `FirebaseListObservable`, implemented consistently like so: 

```typescript

  set(item?: FirebaseOperation, value?: Object): firebase.Promise<void> {
    if (!value){
      throw new Error("The set method requires a defined non-null value. Use remove method to delete instead.");
    }
    // if no item parameter is provided, destructively replace the whole list with value
    if (!item) {
      return this.$ref.ref.set(value);
    }
    return this._checkOperationCases(item, {
      stringCase: () => this.$ref.ref.child(<string>item).set(value),
      firebaseCase: () => (<firebase.database.Reference>item).set(value),
      snapshotCase: () => (<firebase.database.DataSnapshot>item).ref.set(value),
      unwrappedSnapshotCase: () => this.$ref.ref.child((<AFUnwrappedDataSnapshot>item).$key).set(value)
    });
  }
```

What do y'all think?
I could PR if fair game

**Comment by davideast** [comment]
@somombo Thanks for raising this idea! I'm okay with it as long as you follow the implementation of `update` instead of `remove`.

```ts
  set(item: FirebaseOperation, value: Object): firebase.Promise<void> {
    return this._checkOperationCases(item, {
      stringCase: () => this.$ref.ref.child(<string>item).set(value),
      firebaseCase: () => (<firebase.database.Reference>item).set(value),
      snapshotCase: () => (<firebase.database.DataSnapshot>item).ref.set(value),
      unwrappedSnapshotCase: () => this.$ref.ref.child((<AFUnwrappedDataSnapshot>item).$key).set(value)
    });
  }
```

The method should throw if the developer does not provide a `FirebaseOperation`. Otherwise in your code the null check destructively updates the entire list. 

```ts
if (!item) {
  return this.$ref.ref.set(value);
}
```

In that case you'd likely want to use a `FirebaseObjectObservable` since you are now treating treat the list as an object.

**Comment by somombo** [comment]
@davideast that was intensional. 

I have been asking myself after https://github.com/angular/angularfire2/pull/888 is merged, what will a `FirebaseObjectObservable` be able to do that a `FirebaseListObservable` will not? And after looking extensively through the code, this Issue here was what I came up with.

In the interest of DRY familiar API.. My goal is to eventually unify the two in a patch to the rxjs Observable Class so that one could write, in a familiar way: 

```typescript
let mydbObs = Observable.fromDatabase( 
  ref: PathReference, 
  query?: Query | ScalarQuery, 
  selector?: SelectorFunction 
)
```
This would emit lists by default.. and if they happen to be working with an `object`  and not a list... well, the user can select for that, to mould it into precisely what they want.

I was almost getting the unification, but the thing I've proposed in this Issue, was the missing piece. Otherwise, I suppose they'd  have to be two seperate functions still.

Also, this needn't be a replacement to the current AF API we already have, but just an added option that looks more rxjs familiar.

So, in summary, I didn't really intend to model that behavior of `af.list.remove` but of `af.object.set` ... which we seem to have no problem with it being entirely destructive in the way you've pointed out. 

what do you think?

**Comment by davideast** [comment]
I don't agree that having a selector function in a `fromDatabase` method is more familiar to rxjs or DRYer. I think that that it merges responsibilities together when they should in fact be separate. For example the `query` option is not applicable to single objects. 

The goal is to move away from a single API. This provides the benefit that you can tree shake away code you don't need. If you include the `AngularFireDatabase` service, it will bundle both `FirebaseListObservable` and `FirebaseObjectObservable`. However, you still have the option to import their factories and webpack is smart enough to leave behind the code you don't need. 

I'm supportive of this change if it works for a single item in the list. It's okay for the `FirebaseObjectObservable.set()` to destructively update the object since that is the intended behavior. However, it's confusing and dangerous to destructively update an entire list.

**Comment by somombo** [comment]
hi David, I have done as you suggested in #910. Removed scenario where entire list ist replace https://github.com/somombo/angularfire2/tree/%40somombo/add-list-set
I'm ready for PR. Have a couple questions
was hoping to catching you on https://gitter.im/davideast

**Comment by davideast** [comment]
@somombo Send the PR and we can do a code review.

**Comment by jamesdaniels** [comment]
Merged.
